File: src/main/kotlin/org/smarthata/service/MeasureCleaner.java
package org.smarthata.service;

import jakarta.transaction.Transactional;
import lombok.extern.log4j.Log4j2;
import org.smarthata.model.Measure;
import org.smarthata.model.Sensor;
import org.smarthata.repository.MeasureRepository;
import org.springframework.stereotype.Service;

import java.time.LocalDateTime;
import java.time.ZoneId;
import java.util.Date;

@Service
@Log4j2
public class MeasureCleaner {

    private final MeasureRepository measureRepository;

    public MeasureCleaner(MeasureRepository measureRepository) {
        this.measureRepository = measureRepository;
    }

    @Transactional
    public Long cleanPeriod(Sensor sensor, LocalDateTime start, LocalDateTime end) {
        log.debug("Period {} - {}", start, end);

        Date startDate = convertToDateViaInstant(start);
        Date endDate = convertToDateViaInstant(end);

        Long count = measureRepository.countBySensorAndDateBetween(sensor, startDate, endDate);

        if (count <= 1) {
            log.debug("Skip due to small count {}", count);
            return 0L;
        }

        long mid = (endDate.getTime() - startDate.getTime()) / 2;
        Date mediumDate = new Date(startDate.getTime() + mid);

        Double avg = measureRepository.avgValueBySensorAndDateBetween(sensor.id, startDate, endDate);

        Measure measure = new Measure(sensor, avg, mediumDate);
        log.debug("count = {} values avg = {}  medDate = {}, measure = {}", count, avg, mediumDate, measure);

        Long removed = measureRepository.deleteBySensorAndDateBetween(sensor, startDate, endDate);

        measureRepository.save(measure);

        return removed;
    }

    Date convertToDateViaInstant(LocalDateTime dateToConvert) {
        return java.util.Date
                .from(dateToConvert.atZone(ZoneId.systemDefault())
                        .toInstant());
    }

}

File: src/main/kotlin/org/smarthata/service/MeasureCleaner.kt
package org.smarthata.service

import jakarta.transaction.Transactional
import org.slf4j.LoggerFactory
import org.smarthata.model.Measure
import org.smarthata.model.Sensor
import org.smarthata.repository.MeasureRepository
import org.springframework.stereotype.Service
import java.time.LocalDateTime
import java.time.ZoneId
import java.util.Date

@Service
class MeasureCleaner(private val measureRepository: MeasureRepository) {
    private val logger = LoggerFactory.getLogger(this.javaClass)
    @Transactional
    fun cleanPeriod(sensor: Sensor, start: LocalDateTime, end: LocalDateTime): Long {
        logger.debug("Period {} - {}", start, end)
        val startDate = convertToDateViaInstant(start)
        val endDate = convertToDateViaInstant(end)
        val count = measureRepository.countBySensorAndDateBetween(sensor, startDate, endDate)
        if (count <= 1) {
            logger.debug("Skip due to small count {}", count)
            return 0L
        }
        val mid = (endDate.time - startDate.time) / 2
        val mediumDate = Date(startDate.time + mid)
        val avg = measureRepository.avgValueBySensorAndDateBetween(sensor.id, startDate, endDate)
        val measure = Measure(sensor, avg, mediumDate)
        logger.debug("count = {} values avg = {}  medDate = {}, measure = {}", count, avg, mediumDate, measure)
        val removed = measureRepository.deleteBySensorAndDateBetween(sensor, startDate, endDate)
        measureRepository.save(measure)
        return removed
    }

    fun convertToDateViaInstant(dateToConvert: LocalDateTime): Date = Date
        .from(
            dateToConvert.atZone(ZoneId.systemDefault())
                .toInstant()
        )
}